src/components/tl/mlx5/tl_mlx5_team.c

```diff
@@ -169,40 +146,153 @@ static inline ucc_status_t ucc_tl_mlx5_a2a_team_test(ucc_base_team_t *team)
+            return UCC_ERR_NO_RESOURCE;
+        }
+
+        return UCC_OK;
+    }
+
+    ucc_assert(!tl_team->global_sync_in_progress);
+    
+    if (tl_team->mcast_state == TL_MLX5_TEAM_STATE_MCAST_CTX_CHECK &&
+            tl_team->a2a_state == TL_MLX5_TEAM_STATE_ALLTOALL_CTX_CHECK ) {
+        // check if ctx is ready for a2a and mcast
+        tl_team->local_status_array[UCC_TL_MLX5_A2A_STATUS_INDEX] =
+            tl_team->a2a_status.local;
+        tl_team->local_status_array[UCC_TL_MLX5_MCAST_STATUS_INDEX] =
+            (tl_team->local_mcast_ctx_ready) ? UCC_OK : UCC_ERR_NO_RESOURCE;
+        goto initial_sync_post;
+    }
 
     a2a_status = ucc_tl_mlx5_a2a_team_test(team);
     if (a2a_status < 0) {
-        tl_error(team->context->lib, "ALLTOALL tl team: %p creation failed %d",
-                 team, a2a_status);
+        tl_warn(team->context->lib, "ALLTOALL tl team: %p creation failed %d",
+                team, a2a_status);
         tl_team->a2a_state = TL_MLX5_TEAM_STATE_ALLTOALL_NOT_AVAILABLE;
     }
 
     if (tl_team->mcast_state != TL_MLX5_TEAM_STATE_MCAST_NOT_AVAILABLE) {
         mcast_status = ucc_tl_mlx5_mcast_team_test(team);
         if (mcast_status < 0) {
-            tl_error(team->context->lib, "MCAST tl team: %p creation failed %d",
-                     team, mcast_status);
+            tl_warn(team->context->lib, "MCAST tl team: %p creation failed %d",
+                    team, mcast_status);
             tl_team->mcast_state = TL_MLX5_TEAM_STATE_MCAST_NOT_AVAILABLE;
         }
     }
 
-    if (tl_team->mcast_state == TL_MLX5_TEAM_STATE_MCAST_NOT_AVAILABLE &&
-        tl_team->a2a_state == TL_MLX5_TEAM_STATE_ALLTOALL_NOT_AVAILABLE) {
-        tl_error(team->context->lib, "unable to initialize tl team: %p", team);
-        return UCC_ERR_NO_RESOURCE;
-    }
-
     if (UCC_OK != a2a_status || UCC_OK != mcast_status) {
         return UCC_INPROGRESS;
     }
 
-    tl_debug(team->context->lib, "initialized tl team: %p: MCAST component is %s ALLTOALL component is %s",
-            team, (tl_team->mcast_state == TL_MLX5_TEAM_STATE_MCAST_READY)?"ENABLED":"DISABLED",
-            (tl_team->a2a_state == TL_MLX5_TEAM_STATE_ALLTOALL_READY)?"ENABLED":"DISABLED");
-    return UCC_OK;
+    tl_team->local_status_array[UCC_TL_MLX5_A2A_STATUS_INDEX] =
+        (tl_team->a2a_state == TL_MLX5_TEAM_STATE_ALLTOALL_READY) ? UCC_OK : UCC_ERR_NO_RESOURCE;
+    tl_team->local_status_array[UCC_TL_MLX5_MCAST_STATUS_INDEX] =
+        (tl_team->mcast_state == TL_MLX5_TEAM_STATE_MCAST_READY) ? UCC_OK : UCC_ERR_NO_RESOURCE;
+
+    tl_debug(UCC_TL_TEAM_LIB(tl_team),
+             "posting global status, local status: a2a %d mcast %d",
+             (tl_team->a2a_state == TL_MLX5_TEAM_STATE_ALLTOALL_READY),
+             (tl_team->mcast_state == TL_MLX5_TEAM_STATE_MCAST_READY));
+
+initial_sync_post:
+    status = ucc_service_allreduce(
+        core_team, tl_team->local_status_array, tl_team->global_status_array,
```

Comment:
local_status_array is not actually needed, you can just make inplace allreaduce

Reply:
I agree that it is doable, however, it gives us a clear understanding to know our local status of different features in the same array (even after allreduce is done). This can be used for logging or clearing the local resources. Also, it is more readable and it does not consume too much memory.